### PR TITLE
[ROCm] Enable MIOpen properly

### PR DESCRIPTION
**Pull request by iotamudelta**
* Disable MIOpen convolution on double tensors
* MIOpen: set group count in convolution descriptor
* MIOpen: Honor Max Dim (ROCm 222)
* MIOpen: Batchnorm - Allow half/half and half/float, disallow double
* Limit MIOpen batchnorm to same-precision
* Fix maxdim check. (ROCm 246)
* Fix reversed logic in DISABLE_MIOPEN (ROCm 253)
* Export LANG/LC_ALL also for the test step.
* Make tensors contiguous before calling MIOpen batch norm
* Actually pass dilation to MIOpen.
* Do not use MIOpen if there is dilation and the group size is > 1. - This is officially not supported currently.
* Fixes for miopenforward bias call
* Modified init conv descriptor param values and used same value for dilation
* MIOpen: disable transposed convolutions

For attention: @bddppq @ezyang

### CODE REVIEW

**Comment by iotamudelta** [comment]
Caffe2 tests are here: https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/2995/

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -3,6 +3,9 @@
 
 #include "ATen/Config.h"
 
+static const int MIOPEN_DIM_MAX = 4;
+static const bool MIOPEN_ENABLED = getenv("DISABLE_MIOPEN") == NULL;
```
Comment (ezyang) [inline aten/src/ATen/native/Convolution.cpp]: No, we're not going to do this. Please plumb through `miopen_enabled` in the same way `cudnn_enabled` is currently done, if you want to be able to disable it. (But I'm not sure why you would want to disable it.)

**Review by ezyang — CHANGES_REQUESTED** [review]
static initializer `getenv` is a big nono.

File: aten/src/ATen/native/Convolution.cpp (hunk 1 above)
Comment (iotamudelta) [reply aten/src/ATen/native/Convolution.cpp]: It makes debugging easier to be able to switch w/o recompile. But I'll remove it in this PR.

**Review by bddppq — APPROVED** [review]
LGTM.

@ezyang mind taking another look?

**Review by ezyang — APPROVED** [review]
